### PR TITLE
Add test for updateOrCreate

### PR DESCRIPTION
**Pull request by Amir-61**
* Add test for updateOrCreate when id is not autogenerated Id

Connect to https://github.com/strongloop/loopback-connector-db2/issues/38

@loay @superkhau @jannyHou Please review!

CC: @qpresley and @bajtos

### CODE REVIEW

**Comment by Amir-61** [comment]
@slnode test please

**Comment by loay** [comment]
LGTM

File: test/manipulation.test.js [hunk 1]
```diff
@@ -658,6 +658,16 @@ describe('manipulation', function() {
...
+    var ds = getSchema();
+    var Post;
+    before(function(done) {
+      Post = ds.define('Post', {
+        title: { type: String, id: true },
+        content: { type: String },
+        comments: [String],
+      });
+      ds.automigrate('Post', done);
+    });
```
Comment (superkhau) [inline test/manipulation.test.js]: add a space

File: test/manipulation.test.js [hunk 2]
```diff
@@ -658,6 +658,16 @@ describe('manipulation', function() {
   });
 
   describe('updateOrCreate', function() {
+    var ds = getSchema();
+    var Post;
```
Comment (superkhau) [inline test/manipulation.test.js]: add a space here

File: test/manipulation.test.js [hunk 3]
```diff
@@ -658,6 +658,16 @@ describe('manipulation', function() {
...
+    before(function(done) {
```
Comment (superkhau) [inline test/manipulation.test.js]: add a description of what you're doing

```
before('prepare "Post" model', function(done) {
...
```

File: test/manipulation.test.js [hunk 4]
```diff
@@ -658,6 +658,16 @@ describe('manipulation', function() {
...
+      Post = ds.define('Post', {
+        title: { type: String, id: true },
+        content: { type: String },
+        comments: [String],
```
Comment (superkhau) [inline test/manipulation.test.js]: Can this value be String directly?

```
comments: String,
```

File: test/manipulation.test.js [hunk 5]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
         });
     });
 
+    it('update specific instance when PK is not an auto-generated id', function(done) {
```
Comment (superkhau) [inline test/manipulation.test.js]: update -> updates
instance -> instances

File: test/manipulation.test.js [hunk 6]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+      Post.create(
+      [{
+        title: 'postA',
+        content: 'contentA',
+      }, {
+        title: 'postB',
+        content: 'contentB',
+      }],
+      function(err, instance) {
```
Comment (superkhau) [inline test/manipulation.test.js]: We can probably clean up this section a bit.

```
Post.create([
  {title: 'postA', content: 'contentA'},
  {title: 'postB', content: 'contentB'},
], function(err, instance) {
```

File: test/manipulation.test.js [hunk 7]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+        if (err) return done(err);
```
Comment (superkhau) [inline test/manipulation.test.js]: add a space here

File: test/manipulation.test.js [hunk 8]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+        Post.updateOrCreate(
+          { title: 'postA', content: 'newContent' },
+          function(err, updated) {
```
Comment (superkhau) [inline test/manipulation.test.js]: Clean this section up a bit like above:

```
// if this fits on one line (passes linting)
Post.updateOrCreate({title: 'postA', content: 'newContent' }, function(err, instance) {
});

// if not
Post.updateOrCreate({
  title: 'postA', content: 'newContent' 
}, function(err, instance) {
  ...
});
```

File: test/manipulation.test.js [hunk 9]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+            if (err) return done(err);
```
Comment (superkhau) [inline test/manipulation.test.js]: add a space here

File: test/manipulation.test.js [hunk 10]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+            var result = updated.toObject();
```
Comment (superkhau) [inline test/manipulation.test.js]: var result = instance.toObject()`

File: test/manipulation.test.js [hunk 11]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+            result.should.have.properties({
+              title: 'postA',
+              content: 'newContent',
+            });
+            Post.find(function(err, data) {
+              if (err) return done(err);
```
Comment (superkhau) [inline test/manipulation.test.js]: add a space here

File: test/manipulation.test.js [hunk 12]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+        Post.updateOrCreate(
+          { title: 'postA', content: 'newContent' },
+          function(err, updated) {
+            if (err) return done(err);
+            var result = updated.toObject();
+            result.should.have.properties({
+              title: 'postA',
+              content: 'newContent',
+            });
+            Post.find(function(err, data) {
```
Comment (superkhau) [inline test/manipulation.test.js]: data -> posts

File: test/manipulation.test.js [hunk 13]
```diff
@@ -716,6 +726,39 @@ describe('manipulation', function() {
...
+              if (err) return done(err);
+              data.should.have.length(2);
+              data[0].title.should.equal('postA');
+              data[1].title.should.equal('postB');
+              data[0].content.should.equal('newContent');
+              data[1].content.should.equal('contentB');
```
Comment (superkhau) [inline test/manipulation.test.js]: Group post properties together, ie. post A, then post B
```
data[0].title.should.equal('postA');
data[0].content.should.equal('newContent');
data[1].title.should.equal('postB');
data[1].content.should.equal('contentB');
```

**Comment by superkhau** [comment]
@Amir-61 Reviewed, please update and ping again when you're done making the changes.

**Comment by Amir-61** [comment]
@slnode test please

**Comment by superkhau** [comment]
LGTM, but please check why some tests are failing. We should address CI if there are any issues.

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
Thanks @loay @superkhau for reviewing this patch.

>please check why some tests are failing.

@superkhau Yes. Most of the dependencies fail with:

`
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
`

Talked to @rmg and it seems the build machine shutdown mid-job; there are some work being done on cluster that @rmg is aware of. Once done I will run tests again and then land the PR.

Thanks

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@slnode test please

**Comment by Amir-61** [comment]
@bajtos So now based on the new procedure, can we land the patch after getting LGTM from repo owners? Or final LGTM needs to be from you or Raymond? This is a trivial patch, my question is more about more involving changes...

Thanks!

**Comment by loay** [comment]
tests failing are not related to this PR AFAICT

**Comment by rmg** [comment]
> AFAICT, test failures are not related to this PR

The old dependants builder wasn't version aware, so it is building modules that depend on any version of loopback-datasource-juggler instead of only testing ones that depend on 3.x.

That and this PR only adds tests.

**Comment by Amir-61** [comment]
@0candy @supasate @loay 
Sorry I had forgot to back-port this patch to `2.x`; [here](https://github.com/strongloop/loopback-datasource-juggler/pull/1012) is the back-porting patch which I will merge after checking the CI results.

**Pull request merged**